--- lib/awful/wibox.lua.in.old	2010-01-04 23:19:27.286739420 +0200
+++ lib/awful/wibox.lua.in	2010-01-04 23:58:21.813728732 +0200
@@ -144,7 +144,7 @@
 function align(wibox, align, screen)
     local position = get_position(wibox)
     local screen = screen or wibox.screen or 1
-    local area = capi.screen[screen].geometry
+    local area = capi.screen[screen].workarea
 
     if position == "right" then
         if align == "right" then
