--- bash_completion	2008-12-19 23:20:22.000000000 +0200
+++ bash_completion	2009-03-23 02:44:37.212245134 +0200
@@ -206,8 +206,7 @@
 # results in the original argument
 quote_readline()
 {
-	local t="${1//\\/\\\\}"
-	echo \'${t//\'/\'\\\'\'}\' #'# Help vim syntax highlighting
+	echo \'${1//\'/\'\\\'\'}\' #'# Help vim syntax highlighting
 }

 # This function shell-dequotes the argument
@@ -274,7 +274,7 @@ _filedir()
 	local toks=( ) tmp

 	toks=( ${toks[@]-} $(
-		compgen -d -- "$(quote_readline "$cur")" | {
+		compgen -d -- "$(quote "$cur")" | {
 			while read -r tmp; do
 				# TODO: I've removed a "[ -n $tmp ] &&" before,
 				#       and everything works again. If this bug
@@ -290,7 +290,7 @@ _filedir()
 	if [[ "$1" != -d ]]; then
 		xspec=${1:+"!*.$1"}
 		toks=( ${toks[@]-} $(
-			compgen -f -X "$xspec" -- "$(quote_readline "$cur")" | {
+			compgen -f -X "$xspec" -- "$(quote "$cur")" | {
 				while read -r tmp; do
 					[ -n $tmp ] && echo $tmp
 				done
