--- src/kmk/kmkbuiltin/mv.c.orig	2008-09-07 23:35:35.000000000 +0200
+++ src/kmk/kmkbuiltin/mv.c	2008-09-07 23:37:50.000000000 +0200
@@ -48,9 +48,6 @@
 
 #include <sys/types.h>
 #ifndef _MSC_VER
-# ifndef __OS2__
-#  include <sys/acl.h>
-# endif
 # include <sys/param.h>
 # include <sys/time.h>
 # include <sys/wait.h>
@@ -91,6 +88,11 @@
 
 static int	do_move(char *, char *);
 #ifdef CROSS_DEVICE_MOVE
+#ifndef _MSC_VER
+# ifndef __OS2__
+#  include <sys/acl.h>
+# endif
+#endif
 static int	fastcopy(char *, char *, struct stat *);
 static int	copy(char *, char *);
 #endif
