diff -Naurp -Naurp fltk-2.0.x-r6671.old/configure.in fltk-2.0.x-r6671/configure.in
--- fltk-2.0.x-r6671.old/configure.in	2008-08-05 00:53:30.000000000 +0200
+++ fltk-2.0.x-r6671/configure.in	2009-04-11 00:47:36.883659022 +0200
@@ -185,7 +185,7 @@ if test x$enable_shared = xyes; then
 	    ;;
 	Linux* | *BSD*)
 #           DSOCOMMAND="\$(CXX) -Wl,-soname,\$@ \$(LDLIBS) -shared -fPIC $DEBUGFLAG -o"
-            DSOCOMMAND="\$(CXX) \$(LDLIBS) -shared -fPIC $DEBUGFLAG -o"
+            DSOCOMMAND="\$(CXX) \$(LDFLAGS) -shared -fPIC $DEBUGFLAG -o"
 	    if test "$libdir" != "/usr/lib"; then
 		    DSOLINK="-Wl,-rpath,$libdir"
             fi
@@ -611,7 +611,7 @@ case $uname in
 	LIBS="$LIBS -lXext $X_EXTRA_LIBS"
 	CFLAGS="$CFLAGS$X_CFLAGS"
 	CXXFLAGS="$CXXFLAGS$X_CFLAGS"
-	LDFLAGS="$X_LIBS $LDFLAGS"
+	LDFLAGS="$LDFLAGS"
 
 	if test "x$x_includes" != x; then
 	    ac_cpp="$ac_cpp -I$x_includes"
diff -Naurp -Naurp fltk-2.0.x-r6671.old/makeinclude.in fltk-2.0.x-r6671/makeinclude.in
--- fltk-2.0.x-r6671.old/makeinclude.in	2006-04-11 00:06:06.000000000 +0200
+++ fltk-2.0.x-r6671/makeinclude.in	2009-04-11 00:47:04.965513756 +0200
@@ -77,8 +77,8 @@ LOCALIMAGES	= @LOCAL_IMAGETARGETS@
 ZLIBINC		= @ZLIBINC@
 
 # libraries to link with:
-LDLIBS = @LDFLAGS@ @LIBS@
-GLDLIBS = @LDFLAGS@ @GLLIB@ @LIBS@
+LDLIBS = @LIBS@
+GLDLIBS = @GLLIB@ @LIBS@
 LINKFLTK = -L../lib @LINKFLTK@
 LINKFLTKGL = -L../lib @LINKFLTKGL@ @LINKFLTK@
 LINKFLTKFORMS = -L../lib -lfltk2_forms @LINKFLTK@
--- fltk-2.0.x-r6671.old/images/Makefile	2008-08-05 01:03:58.000000000 +0200
+++ fltk-2.0.x-r6671/images/Makefile	2009-04-11 01:13:04.325621387 +0200
@@ -109,7 +109,7 @@ include makedepend
 
 ../lib/$(DSONAME): $(OBJECTS)
 	echo $(DSOCOMMAND) $@ ...
-	$(DSOCOMMAND) $@ $(OBJECTS)
+	$(DSOCOMMAND) $@ $(OBJECTS) $(IMAGELIBS) $(LOCAL_IMAGELIBS) $(LINKFLTK) $(LDLIBS)
 	$(RM) ../lib/$(DSOLINK)
 	$(LN) $(DSONAME) ../lib/$(DSOLINK)
 