--- xorg-server-1.7.3.902/dix/events.c~	2009-12-28 18:42:47.000000000 +0000
+++ xorg-server-1.7.3.902/dix/events.c	2009-12-29 18:34:45.000000000 +0000
@@ -764,7 +764,7 @@
 	    new.y = pSprite->physLimits.y2 - 1;
     if (pSprite->hotShape)
 	ConfineToShape(pDev, pSprite->hotShape, &new.x, &new.y);
-    if ((pScreen != pSprite->hotPhys.pScreen) ||
+    if ((noPanoramiXExtension && (pScreen != pSprite->hotPhys.pScreen)) ||
 	(new.x != pSprite->hotPhys.x) || (new.y != pSprite->hotPhys.y))
     {
 #ifdef PANORAMIX
